### ISSUE TITLE
Subclassing problem in Knora

### ISSUE THREAD

**Issue by sacha-alex** [dhlab-basel/Knora#592]
Dear Knora-Team,

  we’ve ran into a problem during the development of our ontologies, concerning the subclassing of resources and properties. In our case it concerns the property ‘text#hasDescription’ of the resource ‘kuno-raeber/convolute’ (all ontologies’ files are attached). Normally this shouldn’t cause any trouble, but when we want to instantiate a resource of the type "kuno-raeber#Convolute" we getting the message 

  {"status":29,"error":"org.knora.webapi.OntologyConstraintException: Resource class http://www.knora.org/ontology/kuno-raeber#Convolute has no cardinality for property http://www.knora.org/ontology/text#hasDescription"}

In my understanding this message means that the property exists in some way, but its cardinality can’t be resolved. We did some further examinations an noticed that we either have to define the property explicitly in the convolute class again, or just rename it (e.g. from hasDescription to has_Description) to make it work the way we want. The same happens when we're dealing with Convolute defined as a blank node. It works as expected, except there is the same error message when we want to use the property 'hasDescription'  (see file kuno-raeber-ontology-knora.ttl). 
In summary, it seems that sometimes the subclassing mechanism works and sometimes it doesn't.
We would appreciate if you can take a look at this problem. (We tested against Knora's current version and used fuseki. The necessary files are attached.)
Thank you very much in advance.

Best,
Sascha

[Knora.zip](https://github.com/dhlab-basel/Knora/files/1214840/Knora.zip)

**Comment by benjamingeer** [comment]
Hi Sascha, thanks for posting this issue. I haven't been unable to reproduce the problem you're experiencing. I created a branch `wip/kuno-raeber/convolute-test` based on the `develop` branch, in which I just added a test to `org.knora.webapi.e2e.v1.ResourcesV1R2RSpec` that creates a `kuno-raeber:Convolute` with a `text:hasDescription` property. You can run it by checking out that branch and typing the following command at the SBT prompt:

```
> fuseki:test-only *ResourcesV1R2RSpec
[info] ResourcesV1R2RSpec:
[2017-08-10 18:45:32,555] INFO  - HttpTriplestoreConnector - Connection OK: 3270
[info] - Load test data (3 seconds, 475 milliseconds)
[info] The Resources Endpoint
[info] - should create a resource of type kuno-raeber:Convolute with a text:hasDescription property (949 milliseconds)
[info] Run completed in 6 seconds, 444 milliseconds.
[info] Total number of tests run: 2
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 2, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 7 s, completed Aug 10, 2017 6:45:35 PM
```

The test loads your ontologies and admin data, then submits this JSON:

```json
{
    "restype_id": "http://www.knora.org/ontology/kuno-raeber#Convolute",
    "label": "a Convolute",
    "project_id": "http://data.knora.org/projects/kuno-raeber",
    "properties": {
    	"http://www.knora.org/ontology/text#hasTitle": [{"richtext_value":{"utf8str":"test title"}}],
	"http://www.knora.org/ontology/text#hasAlias": [{"richtext_value":{"utf8str":"test alias"}}],
	"http://www.knora.org/ontology/text#hasDescription": [{"richtext_value":{"utf8str":"test description"}}],
	"http://www.knora.org/ontology/text#hasComment": [{"richtext_value":{"utf8str":"test comment"}}]
    }
}
```

For convenience, I commented out the other tests in `ResourcesV1R2RSpec`, but that doesn't affect the result of this test.

**Comment by sacha-alex** [comment]
Dear Ben,
thanks for taking your time to look into my problem.
While your test works fine,I think it just proofs my observation that you have to define a property explicitly (again) to make sure it will be taking into account at a subclassed resource at a subclass level > 1 .
Maybe you can try to run the test with the definition of 'hasDescription' removed, e.g. something like

:Convolute
	a owl:Class;
	rdfs:label "convolute"@en, "Konvolut"@de;
	rdfs:comment """Text convolute of which the elements are authored by Kuno Raeber."""@en;
	rdfs:subClassOf text:Convolute, [
		a owl:Restriction; owl:onProperty text:hasTitle; owl:minCardinality "1"^^xsd:nonNegativeInteger], [
		a owl:Restriction; owl:onProperty text:hasAlias; owl:minCardinality "1"^^xsd:nonNegativeInteger], [
#		a owl:Restriction; owl:onProperty text:hasDescription; owl:minCardinality "1"^^xsd:nonNegativeInteger], [
		a owl:Restriction; owl:onProperty text:hasComment; owl:minCardinality "0"^^xsd:nonNegativeInteger].

Usually I would expect a passing test, too, since the property is already defined in the human ontology (human:Creation). Is this assumption correct, or do I make a mistake here?
Many thanks in advance,
Sascha

**Comment by benjamingeer** [comment]
Hi Sascha, thanks for the clarification. Now I understand what is happening. Properties are indeed inherited recursively, so it is not a problem that subclass level > 1. However, the class `text:Convolute` does not inherit `text:hasDescription` from `human:Creation`, because `text:Convolute` has a property `text:hasConvoluteSizeDescription`, which is a subproperty of `text:hasDescription`. Therefore, in `text:Convolute`, the cardinality for `text:hasConvoluteSizeDescription` overrides (replaces) the one for `text:hasDescription`.

This is described in [the knora-base documentation](http://www.knora.org/documentation/manual/rst/knora-ontologies/knora-base.html#owl-cardinalities):

> cardinalities in the subclass can override cardinalities that would otherwise be inherited from the superclass. Specifically, if a superclass has a cardinality on a property P, and a subclass has a cardinality on a subproperty of P, the subclass’s cardinality overrides the superclass’s cardinality.

Possible solutions:

1. Add a property called `human:hasCreationDescription`, make it a subproperty of `text:hasDescription`, and change the definition of `human:Creation` so it has a cardinality for `human:hasCreationDescription` instead of `text:hasDescription`.  Then when you create a `kuno-raeber:Convolute`, use `human:hasCreationDescription` instead of `text:hasDescription`.
2. Change the definition of `text:hasConvoluteSizeDescription` so that it is not a subproperty of `text:hasDescription`.
3. Give `kuno-raeber:Convolute` a cardinality for `text:hasDescription`.

Personally I think I would find (1) clearest, but it's up to you.
